### ISSUE TITLE
ZeRO-1/2 average_tensor serializes backward compute behind the preceding reduction

### ISSUE THREAD

**Issue by almightychang** [deepspeedai/DeepSpeed#8364]
## Description

ZeRO stage 1/2 with `overlap_comm=true` still serializes the autograd/current stream behind the reduction stream at the start of every `average_tensor()` call:

```python
if self.overlap_comm:
    stream = self.reduction_stream
    if not get_accelerator().resolves_data_dependency():
        for producer_stream in producer_streams:
            stream.wait_stream(producer_stream)
        get_accelerator().current_stream().wait_stream(stream)
```

The producer-to-reduction dependency is required. However, the reverse dependency makes the current stream wait for all previously queued work on `reduction_stream`, including the preceding bucket's collective. When bucket N+1 becomes ready, backward compute is therefore serialized behind bucket N's collective.

This confines overlap to the interval between launching bucket N and preparing bucket N+1. It is especially costly when collective residence is dominated by rank arrival skew rather than data transfer.

## Environment

- Reproduced on DeepSpeed 0.18.7
- Confirmed present by source inspection in 0.19.6 and current `master`
- PyTorch 2.7.0
- 8x H200 on one NVLink node
- ZeRO stage 2
- bf16 communication
- `overlap_comm=true`
- `contiguous_gradients=true`
- `reduce_scatter=true`
- Hugging Face Trainer
- global batch 512, gradient accumulation 2
- 32 samples per rank per micro-batch
- 6,913,191,280 total parameters
- 2,376,274,560 trainable parameters
- 4,536,916,720 frozen parameters
- no CPU or parameter offload

## Profile evidence

The following values are means per rank per micro-batch from eight Kineto traces:

| Metric | Time |
|---|---:|
| Compute kernel union | 489.0 ms |
| NCCL kernel union | 364.0 ms |
| Compute/NCCL overlap | 75.4 ms |
| NCCL pure transfer estimate | 29.9 ms |
| Wait inside collectives | 334.1 ms |
| NCCL outside compute | 288.6 ms |

The traces show the following consistent timeline for non-final gradient buckets:

1. Bucket N launches on `reduction_stream`.
2. Backward compute resumes and prepares bucket N+1.
3. At or immediately before bucket N+1 launches, the compute stream becomes idle while bucket N's NCCL kernel is still resident.
4. Compute resumes only after that preceding collective completes.

Interval attribution assigns NCCL-outside-compute time as follows:

- 212.2 ms (73.5%): the compute stream is stalled on the reduction stream's tail
- 61.6 ms (21.3%): the final bucket is flushed after backward
- 14.8 ms: other collectives

An event replay that keeps NCCL on its own stream, preserves collective order and measured duration, but removes this premature reverse dependency bounds recoverable time at approximately 204 ms per rank per micro-batch. The result is consistent across ranks.

The effect is not explained by transfer bandwidth or SM contention in this profile. NCCL kernels launch with grid `[16, 1, 1]` on 132 SMs. Only about 29.9 ms of 364.0 ms NCCL residence is attributable to transfer. Rank arrival skew dominates: the slowest rank's compute is 652 ms versus a 520 ms rank mean. This makes the reverse dependency particularly visible: one late rank extends the collective, and every rank's compute stream is then held behind that tail at the next bucket enqueue.

## Why this is distinct from existing fixes

- #5606 intentionally added the reverse wait to prevent reuse of an IPG ping-pong buffer before its preceding reduction completed. It also noted a performance gap for short-compute workloads.
- #7805 removed redundant waits after bucket flush and repaired ping-pong index reset, but left the reverse wait inside `average_tensor()`.
- #8061 / #8080 correctly extended the producer-to-reduction dependency to all streams that copied gradients into the IPG bucket. That correctness dependency must remain.

This issue is not proposing removal of producer dependencies. It concerns the location and scope of the reverse dependency that protects buffer reuse.

## Expected behavior

Launching reduction for bucket N should not block unrelated backward compute that produces bucket N+1.

The current/autograd stream should wait for bucket N's reduction only when the same physical IPG ping-pong buffer is about to be reused or overwritten, not whenever `average_tensor()` is called for the next bucket.

## Proposed design direction

Track reduction completion per IPG buffer index:

1. Keep recording all producer streams for each bucket, as implemented by #8080.
2. Make `reduction_stream` wait on those producers before reading the bucket.
3. Record a reduction-complete event for the physical ping-pong buffer after all reduction-stream reads/copies using that buffer have been queued.
4. Before writing into a buffer index again, make the producer/current stream wait on that buffer's completion event.
5. Do not make the current stream wait on the entire reduction stream when merely enqueueing the next bucket.

This should preserve the buffer-reuse safety intended by #5606 and the multi-producer correctness of #8080 while allowing backward compute to overlap the preceding collective.

Special paths that likely need separate handling include oversized parameters, `contiguous_gradients=false`, CPU offload, MoE process groups, and the final bucket flush.

## Validation requirements

A safe change should include:

- regression coverage for #5606's ping-pong buffer reuse race;
- regression coverage for #8061's multi-producer-stream case;
- rapid reuse of both IPG buffer indices;
- gradient, loss, and final-parameter parity against `overlap_comm=false`;
- a CUDA profiler test or reproducible benchmark showing that compute does not stop at the next bucket enqueue while the preceding collective is resident.

I can prepare a draft PR implementing per-buffer completion events if this direction is acceptable to the maintainers.

**Comment by alanhuangyoo** [comment]
I arrived at the same conclusion from the code side, independently, and I have your steps 1–5 already implemented and partly validated. Posting it here rather than opening a PR, since you offered to prepare one and this is your issue — take it, or tell me you would rather I send it, either is fine by me.

What I have, against `cbd303e`:

- `IPGBucket` grows a `reduce_events: dict`, one event per entry of `buffer`.
- `_record_buffer_reduction()` is called at the end of `reduce_ipg_grads`, **inside** the `with stream(reduction_stream)` block, so the event lands after both readers of that entry: the collective in `average_tensor` and the `copy_grads_in_partition` calls that follow it. That is your step 3.
- `_wait_for_buffer_reduction(bucket)` runs immediately before the first write into `buffer[bucket.index]` in `copy_gradients_to_ipg_bucket`, waits on `reduce_events[index]`, and pops it, so only the first write of a bucket fill pays for it. That is your step 4.
- `get_accelerator().current_stream().wait_stream(stream)` is removed from `average_tensor`. That is your step 5.
- Guarded by `overlap_comm and contiguous_gradients and not resolves_data_dependency()`, so the non-contiguous path and accelerators that resolve their own dependencies are untouched.

Since `bucket.index` flips only after `reduce_ipg_grads`, the event waited on is the one from **two** buckets back, which is the "next write to this physical entry" dependency you describe rather than the current bucket's.

### What I can contribute to your validation list

Gradient/loss parity: bit-identical losses across 16 runs — ZeRO stage {1, 2} × gradient accumulation {1, 4} × {master, patched} × {`overlap_comm` on, off}, 2 GPUs, bf16, `reduce_bucket_size=2e6` so the ping-pong indices flip many times per step. Same SHA over the full 40-step loss sequence in every run.

### What I could not measure, and why

I could not reproduce your timing result. My only box is a shared H20 node, and the configuration I care about was too noisy to conclude anything: identical repeats of master + `overlap_comm=true` gave 9.9, 12.1, 20.1, 20.5 and 22.9 ms of `fwd+bwd`, while every other arm sat stably at 16.2–16.5. The variance is larger than the effect, so I have no perf claim to make — your Kineto attribution is the evidence here, not anything I produced. If you have the H200 node, the patch is the part I can supply and the number is the part you can.

### The one hazard I have not cleared

Your "special paths that likely need separate handling" list is where my open question sits, and it is not in the list yet: **ZeRO-2 with gradient accumulation**.

`copy_grads_in_partition` repoints `param.grad` into `grads_in_partition` (`stage_1_and_2.py:1741`) on the reduction stream. For ZeRO-2 the reduce hook fires on every micro-step, not only at the accumulation boundary (`reduce_ready_partitions_and_remove_grads` is gated on `partition_gradients or is_gradient_accumulation_boundary()`), so on the next micro-backward autograd accumulates into a `param.grad` that aliases `grads_in_partition` — a compute-stream write against a reduction-stream write.

The coarse `wait_stream` in `average_tensor` transitively ordered that. A per-buffer event on the IPG buffers does not, because `grads_in_partition` is a different allocation with a different lifetime. My loss-parity runs above include `ga=4` at stage 2 and came out identical, but identical losses in one run is not evidence of race-freedom, so I do not want to claim it is covered.

Two ways out that I can see: extend the event to cover the partition copies as well (record per bucket, wait before the first accumulation that touches a param in that bucket — awkward, because the accumulation point is per-parameter and driven by autograd), or keep a coarse wait on the ZeRO-2 + `ga > 1` path only. I have not built either.

Happy to hand over the patch as-is, or to fold in whichever resolution you and the maintainers prefer.

**Comment by almightychang** [comment]
@alanhuangyoo Thanks for sharing your findings! I’ve already implemented per-buffer completion events in #8366. Would you be interested in reviewing it together, particularly whether it addresses the gradient accumulation concern you raised?

**Comment by almightychang** [comment]
> a compute-stream write against a reduction-stream write

In #8366, `independent_gradient_partition_epilogue()` still calls `get_accelerator().synchronize()` before consuming partition gradients, and clears `param.grad` afterward. The ordinary ZeRO-2 engine path runs this epilogue after every backward, including non-boundary accumulation micro-steps.

My reading is that this completes the partition copies before the next micro-backward, independently of the per-IPG-buffer events. Were you considering a path that bypasses this epilogue synchronization?

**Comment by alanhuangyoo** [comment]
@almightychang You are right and my concern does not hold. I traced the whole chain rather than answering from memory, because I was the one who raised it.

**No, I was not considering a bypass — there is not one.** The three branches all land safely:

```python
# engine.py:3062-3064
# ZeRO stage >= 2 communicates during non gradient accumulation boundaries as well
if self.zero_optimization_partition_gradients():          # stage >= 2
    self.optimizer.overlapping_partition_gradients_reduce_epilogue()
```

That is ungated by the accumulation boundary, and it goes straight to `independent_gradient_partition_epilogue`, so at stage 2 it does run after every micro-backward exactly as you said. Inside it:

```python
if self.overlap_comm:
    if not get_accelerator().resolves_data_dependency():
        get_accelerator().synchronize()
```

- **`overlap_comm=True` on CUDA** — `resolves_data_dependency()` returns `is_synchronized_device()`, which is `False` for CUDA and NPU, so the synchronize runs. This is the only configuration where my hazard exists, and it is ordered.
- **`overlap_comm=False`** — no separate reduction stream, so `copy_grads_in_partition` is not a cross-stream write to begin with. Worth noting this is the *default* at stage 1/2: `overlap_comm` resolves to `stage == ZeroStageEnum.weights` when unset, so a plain ZeRO-2 run has no reduction stream at all.
- **`resolves_data_dependency()=True`** (CPU accelerator, `is_synchronized_device()=True`) — a synchronized device has no asynchrony to order.

So in every configuration there is either a synchronize or no hazard. The coarse `wait_stream` in `average_tensor` was not what was protecting this, and removing it does not expose it. My "identical losses are not evidence of race-freedom" caveat still stands as a statement about evidence, but the reasoning above is the evidence I did not have.

Thank you for pushing back instead of taking it on trust — I would rather have had this found here than have it sit in your PR description as an unresolved caveat.

Happy to review #8366. Two things I will look at specifically, since they are where I would expect a per-buffer event scheme to be thin:

1. the buffer's event being recorded on the right stream and waited before **reuse** of that buffer, not only before consumption of its contents — ping-pong buffers make those two different moments
2. what happens when a bucket is flushed early because it filled mid-parameter, so one parameter's gradient spans two buffers with two events

I will come back on the PR itself rather than here.
